File: scripts/parity_pathologies.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(jsonlite)
  library(lme4)
})

find_repo_root <- function(start = getwd()) {
  d <- normalizePath(start, mustWork = TRUE)
  repeat {
    if (file.exists(file.path(d, "Cargo.toml"))) return(d)
    parent <- dirname(d)
    if (identical(parent, d)) stop("could not find Cargo.toml ancestor")
    d <- parent
  }
}

flag <- function(name, default = NULL) {
  prefix <- paste0("--", name, "=")
  hit <- grep(paste0("^", prefix), commandArgs(trailingOnly = TRUE), value = TRUE)
  if (!length(hit)) return(default)
  sub(prefix, "", hit[[1]], fixed = TRUE)
}

parse_scalar <- function(lines, key) {
  hit <- grep(paste0("^", key, " *="), lines, value = TRUE)
  if (!length(hit)) stop("missing key ", key)
  value <- sub("^[^=]+=", "", hit[[1]])
  value <- trimws(sub("#.*$", "", value))
  if (grepl('^"', value)) return(sub('"$', "", sub('^"', "", value)))
  if (grepl("^[0-9.]+$", value)) return(as.numeric(value))
  value
}

parse_json_value <- function(lines, key) {
  hit <- grep(paste0("^", key, " *="), lines, value = TRUE)
  if (!length(hit)) stop("missing key ", key)
  value <- trimws(sub("^[^=]+=", "", hit[[1]]))
  jsonlite::fromJSON(value)
}

read_spec <- function(path) {
  lines <- readLines(path, warn = FALSE)
  list(
    path = path,
    contract_version = parse_scalar(lines, "contract_version"),
    name = parse_scalar(lines, "name"),
    stratum = parse_scalar(lines, "stratum"),
    group_sizes = as.integer(parse_json_value(lines, "group_sizes")),
    fe_truth = as.numeric(parse_json_value(lines, "fe_truth")),
    re_cov_truth = as.matrix(parse_json_value(lines, "re_cov_truth")),
    family = parse_scalar(lines, "family"),
    link = parse_scalar(lines, "link"),
    seed = as.integer(parse_scalar(lines, "seed")),
    residual_sd = as.numeric(parse_scalar(lines, "residual_sd"))
  )
}

sqrt_psd <- function(sigma) {
  eig <- eigen(sigma, symmetric = TRUE)
  eig$vectors %*% diag(sqrt(pmax(eig$values, 0)), nrow = nrow(sigma)) %*% t(eig$vectors)
}

deterministic_data <- function(spec) {
  n_pred <- length(spec$fe_truth) - 1L
  q <- nrow(spec$re_cov_truth)
  n_slopes <- max(0L, q - 1L)
  sigma_sqrt <- sqrt_psd(spec$re_cov_truth)

  y <- numeric()
  g <- character()
  predictors <- replicate(n_pred, numeric(), simplify = FALSE)
  row_index <- 0L

  for (group_index in seq_along(spec$group_sizes)) {
    z <- c(
      sin(spec$seed + group_index * 1.7),
      cos(spec$seed * 0.5 + group_index * 2.3),
      sin(spec$seed * 0.25 + group_index * 3.1)
    )[seq_len(q)]
    u <- as.numeric(sigma_sqrt %*% z)
    n_g <- spec$group_sizes[[group_index]]
    for (within in seq_len(n_g)) {
      row_index <- row_index + 1L
      centered <- if (n_g <= 1L) 0 else ((within - 1L) - (n_g - 1L) / 2) / (n_g - 1L)
      x <- if (n_pred == 0L) numeric() else vapply(seq_len(n_pred), function(j) {
        centered ^ j + 0.07 * sin((row_index + j) * 0.61)
      }, numeric(1))
      eta <- spec$fe_truth[[1L]]
      if (n_pred > 0L) eta <- eta + sum(spec$fe_truth[-1L] * x)
      if (q >= 1L) eta <- eta + u[[1L]]
      if (n_slopes > 0L) eta <- eta + sum(u[seq_len(n_slopes) + 1L] * x[seq_len(n_slopes)])
      noise_scale <- if (identical(spec$stratum, "reduced-rank")) 0 else 0.1
      y <- c(y, eta + spec$residual_sd * noise_scale * sin(spec$seed + row_index * 0.73))
      g <- c(g, sprintf("g%03d", group_index))
      for (j in seq_len(n_pred)) predictors[[j]] <- c(predictors[[j]], x[[j]])
    }
  }

  out <- data.frame(y = y, g = factor(g))
  for (j in seq_len(n_pred)) out[[paste0("x", j)]] <- predictors[[j]]
  out
}

fit_lme4 <- function(spec) {
  df <- deterministic_data(spec)
  n_pred <- length(spec$fe_truth) - 1L
  q <- nrow(spec$re_cov_truth)
  fe <- if (n_pred == 0L) "1" else paste(c("1", paste0("x", seq_len(n_pred))), collapse = " + ")
  re <- if (q <= 1L) "(1 | g)" else paste0("(1 + ", paste0("x", seq_len(q - 1L), collapse = " + "), " | g)")
  formula_str <- paste("y ~", fe, "+", re)

  warnings_seen <- character()
  t0 <- proc.time()[["elapsed"]]
  fit <- tryCatch(
    withCallingHandlers(
      lme4::lmer(
        stats::as.formula(formula_str),
        data = df,
        REML = TRUE,
        control = lme4::lmerControl(calc.derivs = FALSE)
      ),
      warning = function(w) {
        warnings_seen <<- c(warnings_seen, conditionMessage(w))
        invokeRestart("muffleWarning")
      },
      message = function(m) {
        msg <- sub("\\n$", "", conditionMessage(m))
        if (nzchar(msg)) warnings_seen <<- c(warnings_seen, msg)
        invokeRestart("muffleMessage")
      }
    ),
    error = function(e) e
  )
  runtime_ms <- (proc.time()[["elapsed"]] - t0) * 1000

  if (inherits(fit, "error")) {
    return(list(
      schema_version = "1.0.0", fixture = spec$name, stratum = spec$stratum,
      engine = "lme4::lmer", source = "scripts/parity_pathologies.R",
      status = "error", error = conditionMessage(fit), warnings = I(warnings_seen),
      converged = FALSE, objective = NA_real_, theta = I(numeric()),
      beta = I(numeric()), sigma = NA_real_, loglik = NA_real_, runtime_ms = runtime_ms
    ))
  }

  list(
    schema_version = "1.0.0",
    fixture = spec$name,
    stratum = spec$stratum,
    engine = "lme4::lmer",
    version = paste0("lme4 ", as.character(utils::packageVersion("lme4")), "; R ", getRversion()),
    source = "scripts/parity_pathologies.R",
    status = "ok",
    warnings = I(warnings_seen),
    converged = is.null(fit@optinfo$conv$lme4$messages),
    objective = as.numeric(lme4::REMLcrit(fit)),
    theta = I(as.numeric(lme4::getME(fit, "theta"))),
    beta = I(as.numeric(lme4::fixef(fit))),
    sigma = as.numeric(attr(lme4::VarCorr(fit), "sc")),
    loglik = as.numeric(stats::logLik(fit)),
    runtime_ms = runtime_ms,
    singular = lme4::isSingular(fit)
  )
}

main <- function() {
  repo <- find_repo_root()
  fixture <- flag("fixture", file.path(repo, "tests/fixtures/pathology_corpus/easy.toml"))
  spec <- read_spec(if (grepl("^/", fixture)) fixture else file.path(repo, fixture))
  result <- fit_lme4(spec)
  json <- jsonlite::toJSON(result, auto_unbox = TRUE, pretty = TRUE, na = "null", digits = 17)
  out <- flag("out", NULL)
  if (is.null(out)) {
    cat(json, "\n")
  } else {
    writeLines(json, if (grepl("^/", out)) out else file.path(repo, out))
  }
}

main()

File: scripts/parity_pathologies.jl
#!/usr/bin/env julia

using CategoricalArrays
using DataFrames
using Dates
using LinearAlgebra
using MixedModels
using Printf
using TOML

function repo_root(start=pwd())
    d = abspath(start)
    while true
        isfile(joinpath(d, "Cargo.toml")) && return d
        parent = dirname(d)
        parent == d && error("could not find Cargo.toml ancestor")
        d = parent
    end
end

function flag(name, default=nothing)
    prefix = "--" * name * "="
    for arg in ARGS
        startswith(arg, prefix) || continue
        return split(arg, "=", limit=2)[2]
    end
    return default
end

function sqrt_psd(sigma)
    ev = eigen(Symmetric(Matrix{Float64}(sigma)))
    ev.vectors * Diagonal(sqrt.(max.(ev.values, 0.0))) * ev.vectors'
end

function deterministic_data(spec)
    fe_truth = Float64.(spec["fe_truth"])
    group_sizes = Int.(spec["group_sizes"])
    re_cov_truth = Matrix{Float64}(hcat(spec["re_cov_truth"]...)')
    residual_sd = Float64(spec["residual_sd"])
    seed = Int(spec["seed"])

    n_pred = length(fe_truth) - 1
    q = size(re_cov_truth, 1)
    n_slopes = max(0, q - 1)
    sigma_sqrt = sqrt_psd(re_cov_truth)

    y = Float64[]
    g = String[]
    predictors = [Float64[] for _ in 1:n_pred]
    row_index = 0

    for group_index in eachindex(group_sizes)
        z_pool = [
            sin(seed + group_index * 1.7),
            cos(seed * 0.5 + group_index * 2.3),
            sin(seed * 0.25 + group_index * 3.1),
        ]
        u = sigma_sqrt * z_pool[1:q]
        n_g = group_sizes[group_index]
        for within in 1:n_g
            row_index += 1
            centered = n_g <= 1 ? 0.0 : ((within - 1) - (n_g - 1) / 2) / (n_g - 1)
            x = [centered^j + 0.07 * sin((row_index + j) * 0.61) for j in 1:n_pred]
            eta = fe_truth[1]
            for j in 1:n_pred
                eta += fe_truth[j + 1] * x[j]
            end
            q >= 1 && (eta += u[1])
            for j in 1:n_slopes
                eta += u[j + 1] * x[j]
            end
            noise_scale = spec["stratum"] == "reduced-rank" ? 0.0 : 0.1
            push!(y, eta + residual_sd * noise_scale * sin(seed + row_index * 0.73))
            push!(g, @sprintf("g%03d", group_index))
            for j in 1:n_pred
                push!(predictors[j], x[j])
            end
        end
    end

    df = DataFrame(y=y, g=categorical(g))
    for j in 1:n_pred
        df[!, Symbol("x$j")] = predictors[j]
    end
    return df
end

# Dense REML profiled objective of a fitted LinearMixedModel at θ, evaluated in
# the element type of θ. Mathematically identical to MixedModels' `objective`:
#   logdet(I + ZΛΛ'Z') + logdet(X'M⁻¹X) + (n-p)(1 + log(2π pwrss/(n-p))).
function dense_reml_objective(model, θ::AbstractVector{T}) where {T}
    X = T.(Matrix(model.X))
    y = T.(model.y)
    n, p = size(X)
    blocks = Matrix{T}[]
    for (k, re) in enumerate(model.reterms)
        s = size(re.λ, 1)
        lambda = zeros(T, s, s)
        for (index, (term, row, col)) in enumerate(model.parmap)
            term == k && (lambda[row, col] = θ[index])
        end
        push!(blocks, T.(Matrix(re)) * kron(Matrix{T}(I, length(re.levels), length(re.levels)), lambda))
    end
    ZL = hcat(blocks...)
    C = cholesky(Symmetric(Matrix{T}(I, n, n) + ZL * ZL'))
    XtMiX = X' * (C \ X)
    β = XtMiX \ (X' * (C \ y))
    r = y - X * β
    pwrss = dot(r, C \ r)
    return logdet(C) + logdet(cholesky(Symmetric(XtMiX))) +
           (n - p) * (1 + log(2 * T(pi) * pwrss / (n - p)))
end

# The easy-stratum data are nearly noiseless (σ ≈ 0.02), so the Float64
# profiled objective carries ~1e-9 of cancellation noise against a smallest
# θ-Hessian eigenvalue of ~0.06: any Float64 optimizer stop is determined only
# to ~2e-4 in θ, and Linux and macOS stop at different points. Polish the
# MixedModels optimum by Newton steps on the same objective in 256-bit BigFloat
# (MPFR is correctly rounded, hence platform-independent) and re-evaluate the
# model at the polished θ.
function polish_theta!(model)
    fitted_objective = objective(model)
    theta0 = collect(model.θ)
    theta = setprecision(BigFloat, 256) do
        f(t) = dense_reml_objective(model, t)
        θ = big.(theta0)
        n = length(θ)
        h = big"1e-25"
        unit(k) = (v = zeros(BigFloat, n); v[k] = one(BigFloat); v)
        converged = false
        for _ in 1:50
            g = [(f(θ + h * unit(i)) - f(θ - h * unit(i))) / (2h) for i in 1:n]
            H = [
                (f(θ + h * unit(i) + h * unit(j)) - f(θ + h * unit(i) - h * unit(j)) -
                 f(θ - h * unit(i) + h * unit(j)) + f(θ - h * unit(i) - h * unit(j))) / (4h^2)
                for i in 1:n, j in 1:n
            ]
            isposdef(Symmetric(H)) || error("θ polish: Hessian is not positive definite")
            step = H \ g
            θ -= step
            if maximum(abs, step) < big"1e-40"
                converged = true
                break
            end
        end
        converged || error("θ polish: Newton iteration did not converge")
        Float64.(θ)
    end
    maximum(abs, theta .- theta0) < 1e-2 ||
        error("θ polish moved θ implausibly far from the MixedModels optimum")
    updateL!(MixedModels.setθ!(model, theta))
    objective(model) <= fitted_objective + 1e-8 ||
        error("θ polish: MixedModels objective worsened at the polished θ")
    return model
end

json_string(x) ="\"" * replace(String(x), "\\" => "\\\\", "\"" => "\\\"") * "\""
json_num(x) = isfinite(Float64(x)) ? @sprintf("%.17g", Float64(x)) : "null"
json_array(xs) = "[" * join(json_num.(collect(xs)), ", ") * "]"
json_string_array(xs) = "[" * join(json_string.(collect(xs)), ", ") * "]"

function fit_mmjl(spec)
    df = deterministic_data(spec)
    re_cov_truth = Matrix{Float64}(hcat(spec["re_cov_truth"]...)')
    q = size(re_cov_truth, 1)
    formula = q <= 1 ? @formula(y ~ 1 + x1 + (1 | g)) : @formula(y ~ 1 + x1 + (1 + x1 | g))

    t0 = time()
    status = "ok"
    warning_text = String[]
    model = nothing
    err = nothing
    try
        model = fit(MixedModel, formula, df; REML=true, progress=false)
    catch e
        status = "error"
        err = sprint(showerror, e)
    end
    runtime_ms = (time() - t0) * 1000
    # Only the well-posed easy stratum has an interior optimum to polish; a
    # failed polish is an error, not a silent fallback to the raw fit.
    status == "ok" && spec["stratum"] == "easy" && polish_theta!(model)

    println("{")
    println("  \"schema_version\": \"1.0.0\",")
    println("  \"fixture\": ", json_string(spec["name"]), ",")
    println("  \"stratum\": ", json_string(spec["stratum"]), ",")
    println("  \"engine\": \"MixedModels.jl\",")
    println("  \"version\": ", json_string("MixedModels " * string(pkgversion(MixedModels)) * "; Julia " * string(VERSION)), ",")
    println("  \"source\": \"scripts/parity_pathologies.jl\",")
    println("  \"status\": ", json_string(status), ",")
    println("  \"warnings\": ", json_string_array(warning_text), ",")
    println("  \"converged\": ", status == "ok" ? "true" : "false", ",")
    if status == "ok"
        println("  \"objective\": ", json_num(objective(model)), ",")
        println("  \"theta\": ", json_array(getproperty(model, :theta)), ",")
        println("  \"beta\": ", json_array(coef(model)), ",")
        println("  \"sigma\": ", json_num(getproperty(model, :sigma)), ",")
        println("  \"loglik\": ", json_num(-objective(model) / 2), ",")
    else
        println("  \"error\": ", json_string(err), ",")
        println("  \"objective\": null,")
        println("  \"theta\": [],")
        println("  \"beta\": [],")
        println("  \"sigma\": null,")
        println("  \"loglik\": null,")
    end
    println("  \"runtime_ms\": ", json_num(runtime_ms))
    println("}")
end

function write_provenance_sibling(out_path)
    stem, _ = splitext(basename(out_path))
    prov_path = joinpath(dirname(out_path), stem * ".provenance.json")
    timestamp = string(Dates.now(Dates.UTC)) * "Z"
    mm_ver = string(pkgversion(MixedModels))
    julia_ver = string(VERSION)
    commit = try
        chomp(read(`git rev-parse HEAD`, String))
    catch
        "unknown"
    end
    body = """
    {
      "schema_version": "1.0",
      "generated_at": "$timestamp",
      "crate_commit": "$commit",
      "regenerator": "scripts/parity_pathologies.jl",
      "source_case": null,
      "reference_engine": "MixedModels.jl $mm_ver",
      "notes": "Julia $julia_ver"
    }
    """
    open(prov_path, "w") do io
        write(io, body)
    end
end

function main()
    root = repo_root()
    fixture = flag("fixture", joinpath(root, "tests/fixtures/pathology_corpus/easy.toml"))
    fixture_path = startswith(fixture, "/") ? fixture : joinpath(root, fixture)
    spec = TOML.parsefile(fixture_path)
    out = flag("out", nothing)
    if isnothing(out)
        fit_mmjl(spec)
    else
        out_path = startswith(out, "/") ? out : joinpath(root, out)
        open(out_path, "w") do io
            redirect_stdout(io) do
                fit_mmjl(spec)
            end
        end
        # Pair JSON output with a sibling provenance.json so the
        # fixture_hygiene every_golden_has_provenance_sibling test stays
        # green when this script writes a fresh fixture.
        write_provenance_sibling(out_path)
    end
end

main()
